[16.0][REF] hr_timesheet_begin_end: More extensible, use compute instead of onchange

carmenbianca opened the pull request:
Needed for #691

Internal task: T12630

carmenbianca committed:
[REF] hr_timesheet_begin_end: Make more extensible

The separation of checks into separate methods is needed because I want
to disable one check in another module. This makes the module more
extensible.

The unit_amount_from_start_stop method also makes the module more
extensible.

I have also moved the onchange to the top of the file, according to the
OCA contribution guidelines.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] hr_timesheet_begin_end: Refactor unit_amount to be a compute field

Instead of using onchange, which is less convenient in Odoo 16.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,37 @@ class AccountAnalyticLine(models.Model):
+        readonly=False,
+    )
+
+    @api.model
+    def _add_unit_amount_to_vals_list(self, vals_list):
+        for vals in vals_list:
+            if not vals.get("unit_amount") and (time_start := vals.get("time_start")):
+                vals["unit_amount"] = self.unit_amount_from_start_stop(
+                    time_start, vals.get("time_stop", 0)
+                )
```
I'm not sure this if-statement is valid.

- unit_amount can be 0. Should probably be `vals.get("unit_amount") is not None`
- 0 is a valid `time_start` value which is not truthy…

carmenbianca replied:
no longer valid.

carmenbianca commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,37 @@ class AccountAnalyticLine(models.Model):
+                    time_start, vals.get("time_stop", 0)
+                )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        self._add_unit_amount_to_vals_list(vals_list)
+        return super().create(vals_list)
```
The reason this is here is because—apparently—the default value of a computed field is _not_ its computed value. Or maybe it is, but the constraints get in the way first.

carmenbianca replied:
write above as comment in the code

carmenbianca replied:
After some testing by disabling the restraint, the compute method really isn't triggered after/during creation, and I find this strange.

huguesdk commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -74,13 +118,24 @@ def _check_time_start_stop(self):
+    def _check_time_start_stop(self):
+        self._validate_start_before_stop()
+        self._validate_unit_amount_equal_to_time_diff()
+        self._validate_no_overlap()
```
this should be guarded with:

```python
        if not self.project_id:
            return
```

carmenbianca replied:
added some such guard using a filter.

should this add `project_id` to the constraint? if yes, should all the fields from `_overlap_domain` also be added to the constraint?

```
        return [
            ("id", "!=", self.id),
            ("employee_id", "=", self.employee_id.id),
            ("date", "=", self.date),
            ("time_start", "<", self.time_stop),
            ("time_stop", ">", self.time_start),
        ]
```

i think _technically_ yes, but i am rather wary about constraints with a lot of fields.

huguesdk commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -74,13 +118,24 @@ def _check_time_start_stop(self):
+        self._validate_no_overlap()
+
+    @api.model
+    def _hours_from_start_stop(self, time_start, time_stop):
```
what if one of `time_start` or `time_stop` is `False`?

carmenbianca replied:
I'm not sure this should happen because Floats default to 0 in Odoo.

huguesdk commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,50 @@ class AccountAnalyticLine(models.Model):
+    unit_amount = fields.Float(
+        compute="_compute_unit_amount",
+        store=True,
+        readonly=False,
```
shouldn’t this have an inverse function instead of `readonly=False`? the documentation doesn’t mention `readonly=False` (although the odoo code contains some of those) but states that to allow writing to a computed field, the `inverse` parameter must be used. this would also allow to correctly compute `time_stop` from `time_start` and `unit_amount` (but i don’t know whether this is actually useful).

carmenbianca replied:
not needed, i think. `compute="_method", store=True, readonly=False` is a pattern in odoo to compute whenever the requirements change, but to allow the user to override subsequently.

huguesdk commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,50 @@ class AccountAnalyticLine(models.Model):
+        # that unit_amount is not yet correctly computed in the constraint, and
+        # to skip the constraint if so.
+        self._add_unit_amount_to_vals_list(vals_list)
+        return super().create(vals_list)
```
i dug deeper into the problem and it comes from the fact that the `analytic` module defines `unit_amount` with `default=0.0`. when there is no default, stored computed fields are correctly computed at creation time, before constrains.

the only way to restore this behavior is to set `default=None` on the field. the compute function should then correctly set the field to its default value of `0.0` when needed.

i tried using a callable as the default value, but it must return a value, and any falsy value seems to be converted to 0, thus not resolving the problem.

carmenbianca replied:
i think it's fine to set `default=None` then.

huguesdk commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,50 @@ class AccountAnalyticLine(models.Model):
-            if stop < start:
-                value_to_html(line.time_start, None)
-                value_to_html(line.time_stop, None)
+            line.unit_amount = line.unit_amount_from_start_stop()
```
as this method is also responsible for setting the default value of the field, `unit_amount_from_start_stop()` should ensure that non-timesheet lines (that is, when `project_id` is `False`) get value of `0.0`.

carmenbianca replied:
done, and also added `project_id` to the `api.depends`

carmenbianca committed:
[FIX] hr_timesheet_begin_end: Test uses timesheet lines now

Timesheet lines are characterised by having a project.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] Add hr_timesheet_begin_end to rebel modules

This is a tricky one. _Technically_ there is nothing that makes
hr_timesheet_begin_end incompatible with the other modules here.
However, when the project_id on a timesheet line is changed, unit_amount
is rerecomputed. This has some consequences.

In the sale_timesheet tests, the project_id of a timesheet line
is (sometimes?) recomputed. Subsequently, unit_amount is recomputed and
set to 0 when it _shouldn't_ be. Under a normal flow this wouldn't be a
problem; time_start and time_stop would be populated to correctly
recompute unit_amount. But in the tests, they are not.

One solution is to, in addition to not recomputing the unit_amount of
non-timesheet lines, not recompute the unit_amount of timesheet lines
whose time_start and time_stop are both set to 00:00. However, this
means that resetting these values to 00:00 does not correctly set
unit_amount back to 0, which seems erroneous to me.

Marking this module as a rebel module seems like the best course of
action to me, even though it would be preferable to test this in
conjunction with the other modules.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
I added `hr_timesheet_begin_end` as rebel module. See the commit message.

huguesdk commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,34 @@ class AccountAnalyticLine(models.Model):
+        # raises an exception.
+        #
+        # By setting the default to None, the computation is correctly
+        # triggered. If nothing is computed, None falls back to 0.
```
i think that the comment can be simpler. something like:

```suggestion
        # removing the default of 0.0 to ensure it is computed when not
        # provided. if not computed (because project_id is False), None is
        # automatically converted to 0.
```

huguesdk reviewed (approved):
just a comment about a (too long) comment. otherwise lgtm! thanks @carmenbianca!

huguesdk committed:
[IMP] simplify comment

huguesdk commented:
@OCA/human-resources-maintainers could someone please review or merge this?

huguesdk commented:
`pre-commit` is currently failing because of [erroneous translation strings in the main branch](https://github.com/OCA/timesheet/blob/c2251eb300b91904770e5fe7c48c6f1540884c02/hr_timesheet_sheet/i18n/fr.po#L1074) (detected by `oca-checks-po`). i’ve just fixed them in weblate.

edit: `pre-commit` fixed now.

mathben commented:
I test it and it works!
Before, I had an error when editing begin_date and end_date on tree into form, I got error into method _check_time_start_stop at save
Thanks @carmenbianca
